developer_guides/virtualisation/virtualisation.rst

```diff
@@ -0,0 +1,255 @@
+
+The SOF virtualisation solution also implements a vhost driver.
+
+2. SOF VirtIO design
+********************
+
+The principal point of the SOF VirtIO design is using the IPC API, that is also
+used by the native SOF driver to communicate with the SOF code, running on the
+audio DSP. We use the same IPC structures and we add 3 more VirtIO-specific IPC
+messages, which are never sent to the DSP, so the firmware doesn't need to
+support them. For the most part IPCs from guests are forwarded one-to-one by the
+host driver to the DSP.
+
+2.1. Communication channels
+---------------------------
+
+The goal of the SOF virtualisation is to implement low latency audio
+virtualisation, which means avoiding copying of audio data and stream position
+updates. In this case only one VirtQueue would be required for IPC message
+exchange. However, during the first stage we also transfer data and stream
+position updates via VirtQueues. That means, that currently the SOF uses 3
+VirtQueues: for IPC, audio data, and stream position updates.
+
+TODO: even though currently both playback and capture are supported, they aren't
+supported simultaneously. This should be fixed in future versions.
+
+2.2. Topology
+-------------
+
+When running natively, the SOF uses the ALSA topology API to configure audio
+pipelines on the DSP. The SOF driver reads a platform-specific topology file
+from the filesystem, parses it and sends all the components from it to the DSP
+firmware.
+
+When configured to support virtualised audio, the host topology additionally
+provides connection points for guest audio pipelines. Guest playback pipelines
+connect to mixer and capture pipelines to demultiplexer components. The ALSA /
+ASoC / DAPM subsystems on the host are unaware of any guests, but they should be
+aware of any streaming, taking place on host parts of VM pipelines for
+configuration and power management. That means, e.g. if a guest initialises
+playback, the ASoC subsystem on the host should be aware, that the pipeline,
+connecting the respective mixer and the used codec is currently performing audio
+playback. To achieve this, for each such guest connection point we add a virtual
+PCM as a widget of type "aif_in" or "aif_out" for playback and capture
+respectively, and a virtual 0-size buffer. They are only visible on the host and
+are not sent to the DSP.
+
+When running in a VM, the SOF also uses ALSA topology, but in this case the
+topology should have no relation to the host hardware. The guest audio topology
+should be decided upon by the host system administrator. Therefore we let the
+SOF instance, running on the guest, obtain topology from the host. Such a
+topology should only contain "software components," like PCMs, buffers, software
+volume controls. Those components thus represent partial pipelines, that then
+will be linked to mixer and demux components of the host SOF topology for
+playback and capture respectively.
```

Comment:
no, please don't assume the presence of mixers. For example you could have one guest taking care of BT audio. You would never have any sound coming from the host on that link.